### PR TITLE
Enable an ecs-service to be load balanced, or connect to an external …

### PR DESCRIPTION
**Pull request by npalm**
Added functionality to:
- Connect a target group to the ecs loadbalancer
- Let the module create a target group and listener rule to load balance the ecs service

### CODE REVIEW

**Comment by npalm** [comment]
@mpas @gertjanmaas  pleas can you check the updates I suggest to the module. I focussed to keep the module backwards compatible. If we have breaking changes that we can explain in a few directions for migrations would be fine for me. 

Besides this I planning the following changes
- remove the dependency to aws_vpc_data vor the cidr, since this cause a change, even their isn't a change
- add Fargate as launch type

Since we have now a new module for the load balancer I think we should also drop in the future the load balancer complete from this module.

What do you think?

File: main.tf [hunk 1]
```diff
@@ -28,11 +28,21 @@ resource "aws_ecs_task_definition" "task" {
   task_role_arn         = "${var.task_role_arn}"
 }
 
-resource "aws_ecs_service" "service_alb" {
-  depends_on = ["aws_alb_listener.listener"]
+locals {
+  target_group_arn = "${var.target_group_arn == "" ? element(concat(aws_alb_target_group.target_group.*.arn, list("")), 0) : var.target_group_arn}"
```
Comment (gertjanmaas) [inline main.tf:32]: Why the concat of 2 lists and selecting an element? Can't this just be `aws_alb_target_group.target_group.arn`?

File: variables.tf [hunk 2]
```diff
@@ -213,3 +213,47 @@ variable "ssl_policy" {
   description = "SSL policy applied to an SSL enabled ALB, see https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/elb-security-policy-table.html"
   default     = "ELBSecurityPolicy-TLS-1-2-2017-01"
 }
+
+variable "enable_target_group_connection" {
+  description = "In case `true` a load balanceer is created for the services which will be connected to the target group specified in `target_group_arn`. Creating a load balancer for an ecs service requires a target group with a connected load balancer. To ensure ther right order of creation provide a list of depended arn in `ecs_services_dependencies`"
```
Comment (gertjanmaas) [inline variables.tf]:
```suggestion
  description = "If `true` a load balancer is created for the service which will be connected to the target group specified in `target_group_arn`. Creating a load balancer for an ecs service requires a target group with a connected load balancer. To ensure the right order of creation, provide a list of depended arns in `ecs_services_dependencies`"
```

File: variables.tf [hunk 3]
```diff
@@ -213,3 +213,47 @@ variable "ssl_policy" {
...
+  default     = false
+}
+
+variable "enable_load_balanced" {
+  description = "Creates a services that can be load balanced, a ecs-services, target group and listener rule."
```
Comment (gertjanmaas) [inline variables.tf]: I don't understand the description, can you fix/rephrase it?

File: variables.tf [hunk 4]
```diff
@@ -213,3 +213,47 @@ variable "ssl_policy" {
...
+  default     = false
+}
+
+variable "target_group_arn" {
+  type        = "string"
+  description = "Required for `enable_target_group_connection` provides the target group arn to be connected to the ecs load balancer. Ensure you provide the arn's of th listeners or listeners rule conntected to the target group as `ecs_services_dependencies`."
```
Comment (gertjanmaas) [inline variables.tf]:
```suggestion
  description = "Required for `enable_target_group_connection` provides the target group arn to be connected to the ecs load balancer. Ensure you provide the arns of the listeners or listeners rule conntected to the target group as `ecs_services_dependencies`."
```

File: variables.tf [hunk 5]
```diff
@@ -213,3 +213,47 @@ variable "ssl_policy" {
...
+
+variable "health_check" {
+  type        = "map"
+  description = "Health check for the target group, will overwrite the defaults (merged). Defaults: `protocol=HTTP or HTTPS` depends on `container_ssl`, `path=/`, `matcher=200-399` and `interval=30`."
+  default     = {}
+}
+
+variable "lb_listener_rule_condition" {
+  type        = "map"
+  description = "The condtion fot he LB listener rule created when `enable_load_balanced` is set."
```
Comment (gertjanmaas) [inline variables.tf]:
```suggestion
  description = "The condition for the LB listener rule which is created when `enable_load_balanced` is set."
```

File: variables.tf [hunk 6]
```diff
@@ -213,3 +213,47 @@ variable "ssl_policy" {
...
+
+  default = {
+    field  = "path-pattern"
+    values = ["/*"]
+  }
+}
+
+variable "ecs_services_dependencies" {
+  type        = "list"
+  description = "A list of arn can be provided to which the creation of the ecs services is depended."
```
Comment (gertjanmaas) [inline variables.tf]:
```suggestion
  description = "A list of arns can be provided to which the creation of the ecs service is depended."
```

File: README.md [hunk 7]
```diff
@@ -192,6 +199,7 @@ module "service" {
 | alb_dns_name | DNS address of the load balancer, if created. |
 | alb_route53_dns_name | Route 53 DNS name, if created. |
 | aws_alb_target_group_arn | ARN of the loadbalancer target group. |
+| target_group_arn |  |
```
Comment (gertjanmaas) [inline README.md]:
```suggestion

```

**Comment by gertjanmaas** [comment]
Changes look good, just a couple of things that need some attention. The example needs to be fixed as it's asking for the value of the `dns_name` variable, which according to the `README.md` is disabled.

Both the planned changes are fine by me, Fargate support would be really nice to have.

I agree, once the load balanced ecs service module is tested we can deprecate that functionality from this module.

**Review by gertjanmaas — CHANGES_REQUESTED** [review]
Changes look good, just a couple of things that need some attention. The example needs to be fixed as it's asking for the value of the dns_name variable, which according to the README.md is disabled.

Both the planned changes are fine by me, Fargate support would be really nice to have.

I agree, once the load balanced ecs service module is tested we can deprecate that functionality from this module.

File: main.tf (hunk 1 above)
Comment (npalm) [reply main.tf:32]: No the target group is optional and is controlled by a count, so we have to deal the case their is no target group defined in the module but injected.

File: variables.tf (hunk 3 above)
Comment (npalm) [reply variables.tf]: Updates

File: README.md (hunk 7 above)
Comment (npalm) [reply README.md]: Removed, duplicate

**Comment by npalm** [comment]
@gertjanmaas Thanks for the feedback, comments are processed.

**Review by gertjanmaas — APPROVED** [review]

**Pull request merged**